### PR TITLE
feat: [IGP-99] Refactor `useAttachmentDownload`

### PR DESCRIPTION
**Pull request by adelloste**
## Short description
This PR refactors the `useAttachmentDownload` to replace logic with fp-ts.

## List of changes proposed in this pull request
- `ts/features/messages/hooks/useAttachmentDownload.tsx`: replaced logic with fp-ts

## How to test
- Start dev server
- Perform a login
- Open an PN or MVL message
- Open a message attachment

### CODE REVIEW

File: ts/features/messages/hooks/useAttachmentDownload.tsx [hunk 1]
```diff
@@ -2,17 +2,55 @@ import { useCallback, useEffect, useState } from "react";
...
+) =>
+  TE.tryCatch(
+    () =>
+      ReactNativeBlobUtil.MediaCollection.copyToMediaStore(
+        {
+          name: displayName,
+          parentFolder: "",
+          mimeType: contentType
+        },
+        "Download",
```
Comment (hevelius) [inline ts/features/messages/hooks/useAttachmentDownload.tsx:32]: Is this a placeholder or a string showed to the user? In this case what do you think about use a locale string? `i18n.t("")`

File: ts/features/messages/hooks/useAttachmentDownload.tsx [hunk 2]
```diff
@@ -37,37 +75,28 @@ export const useAttachmentDownload = (
...
+              taskCopyToMediaStore(attachment, path),
+              TE.chain(downloadFilePath =>
+                taskAddCompleteDownload(attachment, downloadFilePath)
+              ),
+              TE.mapLeft(_ =>
+                showToast(i18n.t("messageDetails.attachments.failing.details"))
+              )
+            )
+          )
+        )();
```
Comment (hevelius) [inline ts/features/messages/hooks/useAttachmentDownload.tsx:99]: Is the async/await required? What happen removing `async/await` and using `then/catch` in the pipe instead? What do you think about?

```suggestion
        pipe(
          isAndroid,
          O.fromPredicate(identity),
          TE.fromOption(() => undefined),
          TE.mapLeft(() => ReactNativeBlobUtil.ios.presentOptionsMenu(path)),
          TE.chain(_ =>
            pipe(
              taskCopyToMediaStore(attachment, path),
              TE.chain(downloadFilePath =>
                taskAddCompleteDownload(attachment, downloadFilePath)
              ),
              TE.mapLeft(_ =>
                showToast(i18n.t("messageDetails.attachments.failing.details"))
              )
            )
          )
        )().catch(//...);
```

File: ts/features/messages/hooks/useAttachmentDownload.tsx [hunk 3]
```diff
@@ -95,13 +124,17 @@ export const useAttachmentDownload = (
...
+      downloadPot,
+      pot.toOption,
+      TE.fromOption(() => new Error("The pot is None")),
+      TE.chain(download =>
+        TE.tryCatch(() => RNFS.exists(download.path), E.toError)
+      ),
+      TE.filterOrElse(identity, () => new Error("The file does not exist")),
+      TE.mapLeft(() => dispatch(downloadAttachment.request(attachment))),
+      TE.chainW(() => TE.tryCatch(() => openAttachment(), E.toError))
+    )();
```
Comment (hevelius) [inline ts/features/messages/hooks/useAttachmentDownload.tsx:137]: ditto

**Review by hevelius — CHANGES_REQUESTED** [review]
Tested on iOS simulator with `io-dev-api-server`. Green check. LGTM.

File: ts/features/messages/hooks/useAttachmentDownload.tsx (hunk 1 above)
Comment (adelloste) [reply ts/features/messages/hooks/useAttachmentDownload.tsx:32]: The `Download` is a `MediaType` so we can avoid using `i18n.t("")`

**Review by hevelius — APPROVED** [review]

File: ts/features/messages/hooks/useAttachmentDownload.tsx [hunk 4]
```diff
@@ -37,37 +75,28 @@ export const useAttachmentDownload = (
...
-              description: attachment.displayName,
-              path: downloadFilePath
-            });
-          } catch (e) {
-            showToast(i18n.t("messageDetails.attachments.failing.details"));
-          }
-        }
+        await pipe(
+          isAndroid,
+          O.fromPredicate(identity),
```
Comment (Vangaorth) [inline ts/features/messages/hooks/useAttachmentDownload.tsx:85]: Can we replace `O.fromPredicate` and `TE.fromOption` with just `TE.fromPredicate((_: boolean) => _, () => undefined)`?

File: ts/features/messages/hooks/useAttachmentDownload.tsx [hunk 5]
```diff
@@ -95,13 +124,17 @@ export const useAttachmentDownload = (
...
-    const fileExists = path !== undefined ? await RNFS.exists(path) : false;
-    if (fileExists) {
-      await openAttachment();
-    } else {
-      dispatch(downloadAttachment.request(attachment));
-    }
+    await pipe(
+      downloadPot,
+      pot.toOption,
+      TE.fromOption(() => new Error("The pot is None")),
```
Comment (Vangaorth) [inline ts/features/messages/hooks/useAttachmentDownload.tsx:130]: If I am not wrong, we are not really processing any left branch so what if we replace every `() => new Error` and `E.toError` with `() => undefined`?

**Review by Vangaorth — CHANGES_REQUESTED** [review]

**Review by Vangaorth — APPROVED** [review]

**Comment by Vangaorth** [comment]
LGTM

**Pull request merged**